### ISSUE TITLE
reasoning_effort field silently ignored for step3p7 (and likely other non-Gemma4/Qwen3.5) chat-template models

### ISSUE THREAD

**Issue by True2456** [lmstudio-ai/lmstudio-bug-tracker#2195]
**Which version of LM Studio?**
0.4.19+2

**Which operating system?**
macOS 26.4 (build 25E246), Apple Silicon (M5 Max)

**What is the bug?**

For chat-template-based models whose Jinja template supports an
OpenAI-style `reasoning_effort` variable (e.g. StepFun's `step3p7`
architecture, whose released `chat_template.jinja` has:

```jinja
{%- if reasoning_effort is defined %}
    {{- "Reasoning: " + reasoning_effort + '\n\n' }}
{%- endif %}
```

), sending `"reasoning_effort": "low"` / `"medium"` / `"high"` in an
OpenAI-compatible `/v1/chat/completions` request body has **no effect
whatsoever** -- the field is silently dropped before it ever reaches
`apply_chat_template`. `step3p7`'s own model card documents exactly this
mechanism ("three selectable reasoning levels (low, medium, and high)"),
so this isn't a made-up field -- it's the model's real, published control
surface.

I confirmed the field never reaches the template by asking the model to
echo its own system message verbatim (with a probe user turn) -- no
`Reasoning: <level>` line is present at any `reasoning_effort` value, and
`prompt_tokens` is byte-identical across `low` / `high` / the field being
absent entirely.

I also checked `mlx_engine`'s source (via the local Extensions folder) for
any handling of `reasoning_effort` -- there is none, anywhere. By contrast,
`mlx_engine/tool_protocols.py` and `mlx_engine/tool_runtime.py` have
explicit, hand-wired reasoning support for **Gemma 4** and **Qwen 3.5**
specifically (`gemma4_reasoning_is_open`, `qwen35_reasoning_is_open`,
dedicated guard logits processors tracking `<think>`/`</think>` token
boundaries) -- so LM Studio clearly *has* a per-architecture reasoning
support mechanism, it just hasn't been extended to cover models whose
reasoning control is a plain chat-template variable (`reasoning_effort`)
rather than the guard-based approach used for Gemma4/Qwen3.5.

**To Reproduce**

1. Load any `step3p7`-architecture model (e.g. a local MLX build of
   `stepfun-ai/Step-3.7-Flash`) that supports the documented `low`/
   `medium`/`high` reasoning levels via its `reasoning_effort` chat-template
   variable.
2. Send a request to `/v1/chat/completions` with
   `"reasoning_effort": "high"` (or `"low"`) in the JSON body.
3. Ask the model to repeat its system message verbatim, or compare
   `usage.prompt_tokens` against an identical request with the field
   omitted.
4. Observe: no `Reasoning: <level>` text reaches the model's context,
   `prompt_tokens` is identical regardless of the field's value or
   presence, and reasoning depth/behavior does not change.

**Logs**

No error is logged -- the field is silently ignored rather than rejected,
which is part of why this is easy to miss: there's no error surface telling
the caller their request wasn't honored.

**Suggested fix**

When applying a model's chat template, forward `reasoning_effort` (and any
other OpenAI-compatible sampling/template field not already on a fixed
allowlist) through to `apply_chat_template`'s kwargs if the template
declares it uses that variable -- `mlx_vlm.prompt_utils.apply_chat_template`
already accepts arbitrary `**kwargs` for exactly this purpose, so this may
be as small as forwarding the field through in whatever code path in
`mlx_engine` currently builds the fixed set of template kwargs from the
incoming request body.

**Comment by True2456** [comment]
Follow-up with a sharper root-cause signature, from LM Studio's own server logs (not just an external API round-trip test like the original report).

After the client started correctly sending `"reasoning_effort": "medium"` on every request (confirmed present in the logged outgoing request body), the server log shows this WARN on every single turn:

```
[WARN][step-3.7-p15-vblend-shared8] No valid custom reasoning fields found in model 'truemod/Step-3.7-p15-4bit-vblend-shared8'. Reasoning setting 'medium' cannot be converted to any custom KVs.
```

This confirms the mechanism precisely: LM Studio maintains a fixed per-architecture mapping from a reasoning-level setting to "custom KVs" (presumably the chat-template kwargs it injects, e.g. Jinja `reasoning_effort`), and `step3p7` isn't registered in that mapping -- so the field is silently discarded and the request proceeds without it, exactly as the original report showed via the token-count/system-prompt-echo test.

Non-fatal: generation continues normally immediately after the warning each time (`Running chat completion... Streaming response...`), so this doesn't break anything -- it's just now visible where it was previously silent, once the client actually attempts to send the field on every turn.

Given the WARN's exact wording ("cannot be converted to any custom KVs"), this may be a quick fix if step3p7 just needs adding to whatever table drives that conversion -- happy to test a build if that's useful.

**Comment by snkinard** [comment]
Confirming this issue on the GGUF/llama.cpp path with Qwen3.8 and LM Studio’s official Hub definition.

### Environment

- Server: `llmster v0.0.21+2`
- Model identity: `qwen/qwen3.8-27b@q8_0`
- Official Hub manifest revision: 4
- Context for probe: 8,192
- Endpoint: `/v1/chat/completions`
- Q8_0 SHA-256: `f0fc43cfd802bd87bc7b95f2120553a6407848e3229126b623cc397427857e8a`

Revision 4 defines a Reasoning Effort custom field whose effect is:

```yaml
effects:
  - type: setJinjaVariable
    variable: reasoning_effort
```

I sent the template-native fields directly:

```json
{
  "model": "qwen38-official-medium-chatprobe-nvkbh2",
  "messages": [
    {
      "role": "user",
      "content": "Reply with exactly OK."
    }
  ],
  "chat_template_kwargs": {
    "enable_thinking": true,
    "reasoning_effort": "medium",
    "preserve_thinking": true
  }
}
```

The request completed normally with `finish_reason: "stop"`, final content `OK`, and 44 reasoning tokens. No unsupported-setting, fallback, or ignored-field warning appeared.

However, authoritative formatted-input logging from:

```console
lms log stream --source model --filter input --json
```

showed that the prompt sent to the model contained:

```text
Reasoning effort is set to xhigh. Please think carefully through the task...
```

The pinned Qwen3.8 template defaults an undefined or invalid `reasoning_effort` to `xhigh`. Therefore, the explicit `xhigh` instruction demonstrates that `chat_template_kwargs.reasoning_effort: "medium"` did not reach the Jinja template.

This appears to extend #2195 to:

- GGUF/llama.cpp, not only MLX
- the official LM Studio Hub identity
- a model definition that already declares the correct `setJinjaVariable` mapping
- nested, template-native `chat_template_kwargs`, not only the flat OpenAI-style field

The practical impact is that callers requesting `medium` silently run Qwen3.8 at its much more expensive `xhigh` default. Ideally the endpoint should forward the supplied Jinja variable or reject the field explicitly instead of completing under a different reasoning tier.
